lib/internal/worker/locks.js

```diff
@@ -0,0 +1,526 @@
+    this.channel.unref();
+
+    // Look for existing messages on the channel. There should be some, unless
+    // we are the first thread to set this up.
+    this.readSynchronousMessages();
+    if (!this.state) {
+      // This assertion should be valid because the first Worker instantiation
+      // on the main thread makes sure that locks will be initialized.
+      assert(isMainThread);
+      const sab = new SharedArrayBuffer(16);
+      this.state = {
+        // Use a shared global counter to get ids for the individual
+        // lock requests.
+        lockIdCounter: new BigUint64Array(sab, 0, 1),
+        // The mutex used to ensure that no two threads access the internal
+        // state at the same time. Contains -1 if unlocked, and the locker's
+        // threadId otherwise.
+        mutexOwner: new Int32Array(sab, 8, 1),
+        // Lists of the lock (request)s in their individual states.
+        held: [],
+        pending: [],
+        stolen: [],
+      };
+      this.unlockMutex();
+    }
+  }
+
+  readSynchronousMessages() {
+    let msg;
+    while (msg = receiveMessageOnPort(this.channel)) {
+      this.onmessage(msg.message);
+    }
+  }
+
+  publishState() {
+    this.channel.postMessage({
+      op: 'stateUpdate',
+      state: this.state,
+      [sticky_message_symbol]: true
+    });
+  }
+
+  lockMutex() {
+    // Lock the mutex and load all pending messages in order to update the
+    // internal state.
+    const mutex = this.state.mutexOwner;
+    let owner;
+    while ((owner = Atomics.compareExchange(mutex, 0, -1, threadId)) !== -1) {
+      // We use a timeout to make sure that we still receive messages from
+      // threads that have terminated. It should be very rare to actually
+      // end up blocked on this, because the first other thread to observe
+      // the termination message will use Atomics.notify() to wake us up.
+      while (Atomics.wait(mutex, 0, owner, 100) === 'timed-out') {
+        this.readSynchronousMessages();
+      }
+    }
+    this.readSynchronousMessages();
+  }
+
+  unlockMutex(publish = true) {
+    // Provide the updated internal state to the other threads and
+    // unlock the mutex.
+    if (publish) {
+      this.publishState();
+    }
+    const owner =
+      Atomics.compareExchange(this.state.mutexOwner, 0, threadId, -1);
+    Atomics.notify(this.state.mutexOwner, 0);
+    assert(owner === threadId);
+  }
+
+  request(
+    { resolve, reject },
+    name,
+    shared,
+    ifAvailable,
+    steal) {
+    this.lockMutex();
+
+    // Let request be a new lock request (agent, clientId, origin, name,
+    // mode, promise).
+    const request = {
+      name,
+      shared,
+      clientId: this.clientId,
+      id: Atomics.add(this.state.lockIdCounter, 0, 1n)
+    };
+    let availableResolve, availableReject;
+    const available = new Promise((res, rej) => {
+      availableResolve = res;
+      availableReject = rej;
+    });
+    this.availableMap.set(request.id, {
```

Comment:
Why is this not inside the `new Promise` if `availableResolve` isn't used elsewhere?

Reply:
Because I didn’t think of that :smile: Done!